File: app/src/main/java/com/voipgrid/vialer/dialer/DialerActivity.java
package com.voipgrid.vialer.dialer;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;

import com.voipgrid.vialer.R;
import com.voipgrid.vialer.VialerApplication;
import com.voipgrid.vialer.analytics.AnalyticsHelper;
import com.voipgrid.vialer.calling.Dialer;
import com.voipgrid.vialer.contacts.ContactsSyncTask;
import com.voipgrid.vialer.contacts.SyncUtils;
import com.voipgrid.vialer.t9.ContactsImportProgressUpdater;
import com.voipgrid.vialer.t9.T9Fragment;
import com.voipgrid.vialer.util.ConnectivityHelper;
import com.voipgrid.vialer.util.DialHelper;
import com.voipgrid.vialer.util.JsonStorage;
import com.voipgrid.vialer.util.LoginRequiredActivity;
import com.voipgrid.vialer.util.PhoneNumberUtils;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class DialerActivity extends LoginRequiredActivity implements
        Dialer.Listener,
        T9Fragment.Listener {

    @Inject SharedPreferences mSharedPreferences;
    @Inject AnalyticsHelper mAnalyticsHelper;
    @Inject ConnectivityHelper mConnectivityHelper;
    @Inject JsonStorage mJsonStorage;

    @BindView(R.id.button_call) CallButton callButton;
    @BindView(R.id.bottom) Dialer mDialer;
    @BindView(R.id.top) ViewGroup mTop;
    @BindView(R.id.progress_bar) ContactsProgressBar mProgressBar;
    @BindView(R.id.progress_text) TextView mProgressText;
    @BindView(R.id.contact_processing_container) View mContactsProcessingContainer;
    @BindView(R.id.no_connectivity_container) View noConnectivityContainer;

    private DialHelper dialHelper;
    private T9Fragment mT9Fragment;
    private Thread mContactsProcessingThread;

    /**
     * The key for where the dialer number will be stored
     * when activity returns a result.
     *
     */
    public static final int RESULT_DIALED_NUMBER = 1;

    /**
     * If this extra is present, the dialed number will be returned as a result
     * from the activity rather than making a call. This can be used, for example,
     * when you are trying to prompt the user to provide a transfer destination.
     *
     */
    public static final String EXTRA_RETURN_AS_RESULT = "EXTRA_RETURN_AS_RESULT";

    /**
     * The key in shared preferences where the last dialled number is stored.
     *
     */
    public static final String LAST_DIALED = "last_dialed";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dialer);
        ButterKnife.bind(this);
        VialerApplication.get().component().inject(this);
        dialHelper = DialHelper.fromActivity(this);
        mDialer.setListener(this);
        mT9Fragment = (T9Fragment) getSupportFragmentManager().findFragmentById(R.id.t9_search);
        mT9Fragment.setListener(this);

        preventKeyboardFromBeingDisplayed();
    }

    /**
     * Ensures that a keyboard does not pop-up when pasting into the dialer input field
     *
     */
    private void preventKeyboardFromBeingDisplayed() {
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM,
                WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM);
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (ConnectivityHelper.mWifiKilled) {
            mConnectivityHelper.useWifi(this, true);
            ConnectivityHelper.mWifiKilled = false;
        }

        refreshUi();
    }

    private boolean shouldReturnNumberAsResult() {
        return getIntent().getBooleanExtra(EXTRA_RETURN_AS_RESULT, false);
    }

    public void numberWasChanged(String phoneNumber) {
        mT9Fragment.search(phoneNumber);
    }

    @Override
    public void digitWasPressed(String digit) {

    }

    @Override
    public void exitButtonWasPressed() {

    }

    @Override
    protected void onPause() {
        super.onPause();

        if (mContactsProcessingThread != null && mContactsProcessingThread.isAlive()) {
            mContactsProcessingThread.interrupt();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,String[] permissions, int[] grantResults) {
        if (!allPermissionsGranted(permissions, grantResults)) {
            return;
        }

        if (requestCode == this.getResources().getInteger(R.integer.contact_permission_request_code)) {
            SyncUtils.requestContactSync(this);
            Intent intent = getIntent();
            startActivity(intent);
            finish();
        } else if (requestCode == this.getResources().getInteger(R.integer.microphone_permission_request_code)) {
            dialHelper.callAttemptedNumber();
        }
    }

    /**
     * Initiate an outgoing call by starting CallActivity and pass a SipUri based on the number
     *
     * @param number      number to call
     * @param contactName contact name to display
     */
    public void onCallNumber(String number, String contactName) {
        if (shouldReturnNumberAsResult()) {
            Intent intent = new Intent();
            intent.putExtra("DIALED_NUMBER", number);
            setResult(RESULT_DIALED_NUMBER, intent);
            finish();
            return;
        }

        String phoneNumberToCall = PhoneNumberUtils.format(number);
        if (number.length() < 1) {
            Toast.makeText(this, getString(R.string.dialer_invalid_number), Toast.LENGTH_LONG).show();
        } else {
            dialHelper.callNumber(phoneNumberToCall, contactName);
            mSharedPreferences.edit().putString(LAST_DIALED, number).apply();
        }
    }

    @OnClick(R.id.button_call)
    void onCallButtonClicked() {
        String phoneNumber = mDialer.getNumber();
        if (phoneNumber != null && !phoneNumber.isEmpty()) {
            callButton.setClickable(false);
            onCallNumber(PhoneNumberUtils.format(phoneNumber), null);
        } else {
            // Set last dialed number on call button clicked when number is empty.
            String last_dialed = mSharedPreferences.getString(LAST_DIALED, "");
            mDialer.setNumber(last_dialed);
        }
    }

    @OnClick(R.id.button_dialpad)
    void showKeypad() {
        mDialer.setVisibility(View.VISIBLE);
        findViewById(R.id.button_dialpad).setVisibility(View.INVISIBLE);
        findViewById(R.id.button_call).setVisibility(View.VISIBLE);
    }

    public void refreshUi() {
        callButton.setImageResource(
                shouldReturnNumberAsResult() ?
                        R.drawable.ic_call_transfer :
                        R.drawable.ic_call_white
        );

        if (!hasInternetConnectivity()) {
            noConnectivityContainer.setVisibility(View.VISIBLE);
            mContactsProcessingContainer.setVisibility(View.GONE);
            mT9Fragment.hide();
            mDialer.fadeOut();
            return;
        }

        mDialer.fadeIn();

        if (!ContactsSyncTask.getProgress().isComplete()) {
            if (mContactsProcessingThread == null || !mContactsProcessingThread.isAlive()) {
                mContactsProcessingThread = ContactsImportProgressUpdater.start(this, mProgressText, mProgressBar);
            }
            mContactsProcessingContainer.setVisibility(View.VISIBLE);
            mT9Fragment.hide();
            noConnectivityContainer.setVisibility(View.GONE);
            return;
        }

        mT9Fragment.show();
        noConnectivityContainer.setVisibility(View.GONE);
        mContactsProcessingContainer.setVisibility(View.GONE);
    }


    @Override
    public void onBackPressed() {
        if (isContactsExpanded()) {
            showKeypad();
        } else {
            super.onBackPressed();
        }
    }

    private boolean isContactsExpanded() {
        return mDialer.getVisibility() != View.VISIBLE;
    }

    @Override
    protected void onInternetConnectivityGained() {
        super.onInternetConnectivityGained();
        refreshUi();
    }

    @Override
    protected void onInternetConnectivityLost() {
        super.onInternetConnectivityLost();
        refreshUi();
    }

    @Override
    public void onExpandRequested() {
        if (mDialer.getVisibility() != View.VISIBLE) return;
        mDialer.setVisibility(View.GONE);
        findViewById(R.id.button_dialpad).setVisibility(View.VISIBLE);
        findViewById(R.id.button_call).setVisibility(View.GONE);
    }

    @Override
    public void onContactSelected(String number, String name) {
        onCallNumber(number, name);
    }
}

File: app/src/main/java/com/voipgrid/vialer/dialer/DialerActivity.kt
package com.voipgrid.vialer.dialer

import android.app.AlertDialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.TypedValue
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import butterknife.ButterKnife
import butterknife.OnClick
import com.voipgrid.vialer.R
import com.voipgrid.vialer.User
import com.voipgrid.vialer.VialerApplication
import com.voipgrid.vialer.calling.Dialer
import com.voipgrid.vialer.t9.T9Fragment
import com.voipgrid.vialer.util.ConnectivityHelper
import com.voipgrid.vialer.util.DialHelper
import com.voipgrid.vialer.util.LoginRequiredActivity
import com.voipgrid.vialer.util.PhoneNumberUtils
import kotlinx.android.synthetic.main.activity_dialer.*
import kotlinx.android.synthetic.main.view_dialer.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch


class DialerActivity : LoginRequiredActivity(), Dialer.Listener, T9Fragment.Listener {

    private val dialHelper by lazy { DialHelper.fromActivity(this) }

    private lateinit var t9Fragment: T9Fragment

    private val isContactsExpanded: Boolean
        get() = dialer.visibility != View.VISIBLE

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dialer)
        ButterKnife.bind(this)
        VialerApplication.get().component().inject(this)
        dialer.setListener(this)
        t9Fragment = supportFragmentManager.findFragmentById(R.id.t9_search) as T9Fragment
        t9Fragment.listener = this
        preventKeyboardFromBeingDisplayed()
        handleIntents()
    }

    /**
     * Handles when this activity has been launched from outside the app, for example
     * a user clicking on a phone number link in a web browser.
     *
     */
    private fun handleIntents() {
        val action = intent.action
        val data = intent.dataString ?: return

        logger.i("Received intent to automatically launch dialer: " + intent.action + ", with data: " + data)

        val number = PhoneNumberUtils.format(data)

        when(action) {
            Intent.ACTION_CALL -> onCallNumber(number, null)
            Intent.ACTION_DIAL -> dialer.number = number
        }
    }

    /**
     * Ensures that a keyboard does not pop-up when pasting into the dialer input field
     *
     */
    private fun preventKeyboardFromBeingDisplayed() = window.setFlags(WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM, WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM)

    override fun onResume() {
        super.onResume()

        if (ConnectivityHelper.mWifiKilled) {
            connectivityHelper.useWifi(this, true)
            ConnectivityHelper.mWifiKilled = false
        }

        refreshUi()

        // If we have ever disabled the call button, we don't want the user to return
        // to this activity so just close.
        if (!button_call.isClickable) {
            finish()
        }
    }

    private fun shouldReturnNumberAsResult(): Boolean = intent.getBooleanExtra(EXTRA_RETURN_AS_RESULT, false)

    override fun numberWasChanged(phoneNumber: String) {
        t9Fragment.search(phoneNumber)
    }

    override fun digitWasPressed(digit: String) {

    }

    override fun exitButtonWasPressed() {

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (!allPermissionsGranted(permissions, grantResults)) {
            return
        }

        when(requestCode) {
            this.resources.getInteger(R.integer.contact_permission_request_code) -> {
                startActivity(intent)
                finish()
            }
            this.resources.getInteger(R.integer.microphone_permission_request_code) -> dialHelper.callAttemptedNumber()
        }
    }

    /**
     * Initiate an outgoing call by starting CallActivity and pass a SipUri based on the number
     *
     * @param number      number to call
     * @param contactName contact name to display
     */
    private fun onCallNumber(number: String, contactName: String?) = GlobalScope.launch(Dispatchers.Main) {
        if (isConnectedToNetwork() && !internetConnectivity.canGuaranteeAnInternetConnection()) {
            handleNoInternetAccess()
            return@launch
        }

        if (shouldReturnNumberAsResult()) {
            setResult(RESULT_DIALED_NUMBER, Intent().apply { putExtra("DIALED_NUMBER", number) })
            finish()
            return@launch
        }

        val phoneNumberToCall = PhoneNumberUtils.format(number)

        when(number.isEmpty()) {
            true -> Toast.makeText(this@DialerActivity, getString(R.string.dialer_invalid_number), Toast.LENGTH_LONG).show()
            false -> {
                dialHelper.callNumber(phoneNumberToCall, contactName)
                User.internal.lastDialledNumber = phoneNumberToCall
            }
        }
    }

    private fun handleNoInternetAccess() {
        button_call.isClickable = true

        AlertDialog.Builder(this).setTitle(R.string.dialer_no_internet_connectivity)
                .setMessage(R.string.dialer_no_internet_connectivity_subtext)
                .setCancelable(false)
                .setPositiveButton(this.getString(R.string.ok)) { dialog, _ ->
                    dialog.dismiss()
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) startActivityForResult(Intent(Settings.Panel.ACTION_INTERNET_CONNECTIVITY), 1)
                }
                .show()
    }

    @OnClick(R.id.button_call)
    internal fun onCallButtonClicked() {
        when {
            dialer.number.isNotBlank() -> {
                button_call.isClickable = false
                onCallNumber(PhoneNumberUtils.format(dialer.number), null)
            }
            dialer.number.isBlank() -> dialer.number = User.internal.lastDialledNumber
        }
    }

    @OnClick(R.id.button_dialpad)
    internal fun showKeypad() {
        dialer.visibility = View.VISIBLE
        findViewById<View>(R.id.button_dialpad).visibility = View.INVISIBLE
        findViewById<View>(R.id.button_call).visibility = View.VISIBLE
        t9_search.view?.layoutParams = ConstraintLayout.LayoutParams(
                ConstraintLayout.LayoutParams.MATCH_PARENT,
                resources.getDimensionPixelSize(R.dimen.t9_height)
                )
    }

    private fun refreshUi() {
        button_call.setImageResource(if (shouldReturnNumberAsResult()) R.drawable.ic_call_transfer else R.drawable.ic_call_white)

        dialer.fadeIn()
        t9Fragment.show()
        no_connectivity_container.visibility = View.GONE

        if (!isConnectedToNetwork()) {
            updateUiForNoInternetConnectivity()
            return
        }
    }

    /**
     * Hides various elements to display that there is no internet connectivity.
     *
     */
    private fun updateUiForNoInternetConnectivity() {
        no_connectivity_container.visibility = View.VISIBLE
        t9Fragment.hide()
        dialer.fadeOut()
    }

    override fun onBackPressed() {
        when(isContactsExpanded) {
            true -> showKeypad()
            false -> super.onBackPressed()
        }
    }

    override fun onInternetConnectivityGained() {
        super.onInternetConnectivityGained()
        refreshUi()
    }

    override fun onInternetConnectivityLost() {
        super.onInternetConnectivityLost()
        refreshUi()
    }

    override fun onExpandRequested() {
        if (dialer.visibility != View.VISIBLE) return
        dialer.visibility = View.GONE
        t9_search.view?.layoutParams = ConstraintLayout.LayoutParams(ConstraintLayout.LayoutParams.MATCH_PARENT, ConstraintLayout.LayoutParams.MATCH_PARENT)

        findViewById<View>(R.id.button_dialpad).visibility = View.VISIBLE
        findViewById<View>(R.id.button_call).visibility = View.GONE
    }

    override fun onContactSelected(number: String, name: String) {
        onCallNumber(number, name)
    }

    companion object {

        /**
         * The key for where the dialer number will be stored
         * when activity returns a result.
         *
         */
        const val RESULT_DIALED_NUMBER = 1

        /**
         * If this extra is present, the dialed number will be returned as a result
         * from the activity rather than making a call. This can be used, for example,
         * when you are trying to prompt the user to provide a transfer destination.
         *
         */
        const val EXTRA_RETURN_AS_RESULT = "EXTRA_RETURN_AS_RESULT"

        /**
         * The key in shared preferences where the last dialled number is stored.
         *
         */
        const val LAST_DIALED = "last_dialed"
    }
}